### PR TITLE
[WIP] feat(Button): add "as" prop to allow custom tags

### LINKED ISSUE #161: Button should take a Component prop
Right now, the Button component will only render a `button`. However it's very common for a user to want to implement button styles on an `<a>` element or even a custom link component (such as if they are using client-side routing). So the Button component should be modified to accept an optional `Component` prop (probably of type `ReactNode`) to support that use case. It should still default to `button` if no `Component` prop is defined.

### PR DESCRIPTION
**Pull request by christopherhuii**
# Summary
Currently, the `Button` component only allows it to be a HTML `button`. This PR adds an `as` prop to allow tags and props to be configurable.

For example:
```
// Button
<Button type="submit" onClick={() => {}}>
    Submit
</Button>

// Anchor Tag - Needs href and can accept other props for further configuration
<Button
    component="a"
    href="https://truss.works"
    target="_blank"
    rel="noopener noreferrer"
>
    Visit Truss
</Button>

// React Router Link - Needs "to" prop for navigation
<Button component={Link} to="/login">Login</Button>
```

## Related Issues or PRs

Closes #161 

## How To Test

View code, tests, and stories

### CODE REVIEW

File: src/components/Button/Button.stories.tsx [hunk 1]
```diff
@@ -75,3 +75,30 @@ export const customClass = (): React.ReactElement => (
     Click Me
   </Button>
 )
+
+export const asAnchorTag = (): React.ReactElement => (
+  <Button
```
Comment (haworku) [inline src/components/Button/Button.stories.tsx:80]: This seems like an example of where using  `as` props are any `string` could be an issue- there's no way to verify, for example, that they are passing types that are valid for the anchor tag.   Or that they even pass a valid html element tag name as a string..

File: src/components/Button/Button.tsx [hunk 2]
```diff
@@ -22,14 +21,14 @@ interface ButtonProps {
   small?: boolean
   icon?: boolean
   unstyled?: boolean
+  [x: string]: any
```
[the remark this answers is not in the document]
Comment (christopherhuii) [reply src/components/Button/Button.tsx:24]: I'm not sure if this line was the best way to do this.

I think we are losing type safety. It allows the user to add whatever prop they need. In the case of anchor tags, `target` and `rel` are fairly common to add. For accessibility purposes, `aria-*` can be added. Users may also want to add `data-*`. I added it to accommodate for the various use cases a user may have. I'm not sure what the best way is to move this.

[the remark this answers is not in the document]
Comment (ahobson) [reply src/components/Button/Button.tsx:24]: I think my exploration of how to extend props to allow valid property names in #267 is relevant to this.

**Comment by haworku** [comment]
Changed status to WIP since what happens with #263 will influence the approach here.

**Comment by suzubara** [comment]
@christopherhuii @haworku @ahobson think we can close this in favor of the Link component work?

**Comment by haworku** [comment]
@suzubara Oh, are you are satisfied with `Link`  to meet #161? If so, lets close both!!!  Otherwise I was thinking we might still adapt Button to pass its styles to custom components ( but using the approach Link that also preserves some type safety).

**Pull request closed without merging**